client/rpc-spec-v2/src/archive/archive.rs

```diff
@@ -0,0 +1,304 @@
+) -> Vec<Block::Hash>
+where
+	Block: BlockT + 'static,
+	BE: Backend<Block> + 'static,
+{
+	let mut result = Vec::new();
+	let mut next_hash = Vec::new();
+	next_hash.push(parent);
+
+	while let Some(parent) = next_hash.pop() {
+		if parent.number == target_height {
+			result.push(parent.hash);
+			continue
+		}
+
+		let Ok(blocks) = backend.blockchain().children(parent.hash) else {
+			continue
+		};
+
+		let child_number = parent.number.saturating_add(One::one());
+		for child_hash in blocks {
+			next_hash.push(HashAndNumber { number: child_number, hash: child_hash });
+		}
+	}
+
+	result
+}
+
+#[async_trait]
+impl<BE, Block: BlockT, Client> ArchiveApiServer<Block::Hash> for Archive<BE, Block, Client>
+where
+	Block: BlockT + 'static,
+	Block::Header: Unpin,
+	BE: Backend<Block> + 'static,
+	Client: BlockBackend<Block>
+		+ ExecutorProvider<Block>
+		+ HeaderBackend<Block>
+		+ HeaderMetadata<Block, Error = BlockChainError>
+		+ BlockchainEvents<Block>
+		+ StorageProvider<Block, BE>
+		+ 'static,
+{
+	fn archive_unstable_body(
+		&self,
+		mut sink: SubscriptionSink,
+		hash: Block::Hash,
+		_network_config: Option<NetworkConfig>,
+	) -> SubscriptionResult {
+		let client = self.client.clone();
+
+		let fut = async move {
+			let event = match client.block(&BlockId::Hash(hash)) {
+				Ok(Some(signed_block)) => {
+					let extrinsics = signed_block.block.extrinsics();
+					let result = format!("0x{}", HexDisplay::from(&extrinsics.encode()));
+					ArchiveEvent::Done(ArchiveResult { result })
+				},
+				Ok(None) => ArchiveEvent::Inaccessible,
+				Err(error) => ArchiveEvent::Error(ErrorEvent { error: error.to_string() }),
+			};
+			let _ = sink.send(&event);
+		};
+
+		self.executor.spawn("substrate-rpc-subscription", Some("rpc"), fut.boxed());
+		Ok(())
+	}
+
+	fn archive_unstable_genesis_hash(&self) -> RpcResult<String> {
+		Ok(self.genesis_hash.clone())
+	}
+
+	fn archive_unstable_hash_by_height(
+		&self,
+		mut sink: SubscriptionSink,
+		height: String,
+		_network_config: Option<NetworkConfig>,
+	) -> SubscriptionResult {
+		let height_str = height.trim_start_matches("0x");
+		let Ok(height_num) = u32::from_str_radix(&height_str, 16) else {
+			let _ = sink.reject(ArchiveRpcError::InvalidParam(height));
+			return Ok(())
+		};
+
+		let client = self.client.clone();
+		let backend = self.backend.clone();
+
+		let fut = async move {
+			let finalized_number = client.info().finalized_number;
+
+			// If the height has been finalized, return only the finalized block.
+			if finalized_number >= height_num.into() {
+				let result = if let Ok(Some(hash)) = client.block_hash(height_num.into()) {
+					vec![hash]
+				} else {
+					// The block hash should have existed in the database. However,
+					// it may be possible that it was pruned.
+					vec![]
+				};
+
+				let _ = sink.send(&ArchiveEvent::Done(ArchiveResult { result }));
+				return
+			}
+
+			let finalized_hash = client.info().finalized_hash;
```

Comment:
what's the point of returning finalized blocks here if that's not what the client has asked for? shouldn't we return an error instead saying that "the block with the given height has not been finalized yet"?

Reply:
Thanks for having a look at this!
I seem to have lost this in notifications, the `archive` class is not really necessary for light-client integrations and there are a few things to decide before moving forward with this:
https://github.com/paritytech/json-rpc-interface-spec/pull/26, https://github.com/paritytech/json-rpc-interface-spec/issues/27, https://github.com/paritytech/json-rpc-interface-spec/issues/30, https://github.com/paritytech/json-rpc-interface-spec/pull/29

Regarding this, we are also exploring adding another method `archive_finalizedBlock` to make sure users are aware they can make queries to ancestor blocks of finalized. IIRC, I placed this to make sure they are not making a call to any children of the finalized.